### PR TITLE
fix(launchd): improve `StartCalendarInterval`

### PR DESCRIPTION
**Pull request by tmillr**
Stricter launchd -> StartCalendarInterval type:

- Verify that the integers passed to `Minute`, `Hour`, etc. are within range.

- When provided, the value for StartCalendarInterval must be a non-empty list of calendar intervals and must not contain duplicates entries (throw an error otherwise).

- For increased flexibility and backwards-compatibility, allow an attrset to be passed as well (which will be type-checked and is functionally equivalent to passing a singleton list). Allowing an attrset or list is precisely in-line with what `launchd.plist(5)` accepts for StartCalendarInterval.

Migrate `nix.gc.interval` and `nix.optimise.interval` over to use this new type, and update their defaults to run weekly instead of daily.

Create `modules/launchd/types.nix` file for easier/modular use of launchd types needed in multiple files.

Documentation:

- Update and improve wording/documentation of launchd's `StartCalendarInterval`.

- Improve wording/documentation of `nix.gc.interval` and `nix.optimise.interval` ("time interval" can be misleading as it's actually a "calendar interval"; e.g. `{ Hour = 3; Minute = 15;}` runs daily, not every 3.25 hours).

### CODE REVIEW

File: modules/services/nix-gc/default.nix
Comment (emilazy) [inline modules/services/nix-gc/default.nix:42]: Thank you, this is great! Would you mind making a `modules/launchd/types.nix` and exporting the type without the `nullOr` from it, so that it can be used directly in both `launchd.nix` and `nix-gc/default.nix` instead of having to do this hack?

**Review by emilazy — CHANGES_REQUESTED** [review]

**Comment by steveej** [comment]
hey, since #611 was dropped in favor of this i'm interested in getting this PR merged.

@tmillr thanks for the work on this. do you have bandwidth to see this through or would you prefer someone else continues here?

**Comment by tmillr** [comment]
I'm not sure when I'll be able to finish this. I'm fine with someone else working on this pr or finishing it. Thanks

**Comment by tmillr** [comment]
I've had more time lately and so I've decided I'd try to get this finished once and for all (hopefully within the next day or 2). Sorry for the delay.

**Comment by tmillr** [comment]
## Updates

### Changes

- Calendar interval must be a list (when provided). I did this because allowing an attrset as well ruined the documentation.
- Create `launchd/types.nix` file
- Update option documentation
- Add type-check for duplicate interval specs (i.e. attrsets) within the provided list
- Saner default for gc interval (i.e. the default is now to run weekly instead of daily)

### Questions/Concerns/Etc.

- An error message produced by the new type-checking might not be very detailed/helpful? (e.g. an error due to finding a duplicate interval spec within the list, or, an error due to the list being empty)
- A warning should probably be emitted if an empty attrset is found in the list? (because I am pretty sure that this will result in a job which runs every minute). I don't think Nix recommends emitting warnings within the type-check itself though, so I imagine that this would have to be done anywhere/everywhere `StartCalendarInterval` is used?
- Similarly, there's no check for if there's a list containing an empty attrset in it along with others which aren't empty (wacky edge-case, probably not a big deal? considering the above is implemented at some point)

**Comment by emilazy** [comment]
Thanks for all your work on this! I will try to take a look in the next couple days.

**Comment by steveej** [comment]
@emilazy if i'm not mistaken this is waiting for your review.

**Comment by tmillr** [comment]
Hi @emilazy. I think it's good to go now if you'd like to review it again. I've addressed the merge conflicts (which mainly had to do with the recent removal of `lib.mdDoc` in def1e23). All checks are passing and I've also done some manual testing locally. Let me know if anything else needs to be changed. Thanks

**Comment by emilazy** [comment]
Thank you for all the work you’ve put into this, and I’m sorry for my prolonged absence! Catching up now and will try to review this in the next few days. Please do feel free to ping me if it slips through the cracks.

**Review by emilazy — APPROVED** [review]
Looks great, thank you for sticking with this for so long! I’m wondering what the failure case would be on the launchd end if a duplicate was inadvertently included? The attention to detail to getting a good error message on the nix-darwin end for that case is much appreciated, but I wonder how much we’re getting for the implementation complexity. I’m going to land this now, though, and we can worry about that tradeoff later. Thank you again for all your work on this.

**Pull request merged**
